### ISSUE TITLE
Bug: macOS treating CFData as a string loses information

### ISSUE THREAD

**Issue by directionless** [osquery/osquery#6642]
I was digging at the nvram table, and I noticed that we pass CFData to `stringFromCFData`, which does:
```
    if (isprint(byte)) {
      result << byte;
    } else if (range.length > 1 && buffer[i] == 0) {
      result << ' ';
    } else {
      result << '%' << std::setfill('0') << std::setw(2) << std::hex
             << (int)byte;
    }
```

This makes it hard to reconstruct the binary plists which apple likes to embed in places.

One solution would be to make a `base64FromCFData` and call that instead. Perhaps another might be to have a function to convert that back to the correct binary data.

**Comment by directionless** [comment]
I tried doing the "obvious"
```
  std::string base64FromCFData(const CFDataRef& cf_data) {
    auto length = CFDataGetLength(cf_data);

    CFRange range = CFRangeMake(0, length);

  char* buffer = (char*)malloc(length);
  if (buffer == nullptr) {
    return "";
  }
  memset(buffer, 0, length + 1 );

  CFDataGetBytes(cf_data, range, (UInt8*)buffer);

  return base64::encode(std::string(buffer));
```

But, the resultant base64 is wrong, and too short.

**Comment by theopolis** [comment]
I agree that base64 would be a better approach. @directionless there are a few issues in the example code you posted. First you are `memset`ing `length + 1` but you have only allocated `length`, this is a over-write by 1 byte. 

Second you are using the default `std::string` constructor from a `char*`, this will create a string of bytes up until the first NUL character. Instead you should use the constructor that takes a specific size as a parameter.

**Comment by theopolis** [comment]
I think a unit test with some example data would be helpful. I can then work with you on a PR to get the correct implementation.
